File: app/dados.js
/* Acesso aos arquivos de dados do repositório.

   A página NÃO fala com a internet. Tudo que ela lê está versionado aqui:

     dados/catalogo/ativos.json       o que existe para simular
     dados/premissas/premissas.json   palpites e regras fiscais
     dados/mercado/indicadores.json   retrato do mercado + manifesto das séries e fundos
     dados/mercado/series/*.json      histórico mensal dos indicadores
     dados/mercado/fundos/*.json      histórico mensal de cada FII/ETF (dividend
                                       yield, valorização e/ou preço, ver docs/04)

   Os dois últimos são gerados pelo coletor (coletor/atualizar.py). Nada aqui
   sabe quais indicadores ou séries existem: a lista vem do próprio retrato,
   com rótulo e unidade. Indicador novo é edição de um lugar só, no coletor.

   Consequência a encarar: o dado tem idade. Ela viaja junto com o número, e a
   página mostra quando o retrato ficou velho demais. */

const CAMINHOS = {
  ativos: 'dados/catalogo/ativos.json',
  premissas: 'dados/premissas/premissas.json',
  mercado: 'dados/mercado/indicadores.json',
};
const DIR_MERCADO = 'dados/mercado/';
const DIR_FUNDOS = 'dados/mercado/fundos/';

/** Acima disso a página avisa que o retrato está velho. */
export const DIAS_ATE_VENCER = 10;

const memoria = new Map();

async function lerJson(caminho) {
  if (memoria.has(caminho)) return memoria.get(caminho);
  const promessa = fetch(caminho, { cache: 'no-cache' }).then((r) => {
    if (!r.ok) throw new Error(`Não consegui ler ${caminho} (HTTP ${r.status})`);
    return r.json();
  });
  memoria.set(caminho, promessa);
  return promessa;
}

export async function catalogo() {
  return lerJson(CAMINHOS.ativos);
}

export async function ativosPorId() {
  const { ativos } = await catalogo();
  return Object.fromEntries(ativos.map((a) => [a.id, a]));
}

export async function premissas() {
  return lerJson(CAMINHOS.premissas);
}

async function retratoMercado() {
  try {
    return await lerJson(CAMINHOS.mercado);
  } catch (erro) {
    return { _erro: erro.message, campos: {}, series: [] };
  }
}

function diasDesde(iso) {
  if (!iso) return null;
  const quando = new Date(iso);
  if (Number.isNaN(quando.getTime())) return null;
  return Math.floor((Date.now() - quando.getTime()) / 86400000);
}

/** Indicadores prontos para o motor, com a procedência de cada campo.

    O motor lê os valores por nome (`indicadores.cdi_aa`); a página lê
    `_meta` para montar os cartões sem saber de antemão quais existem. */
export async function indicadores() {
  const premissasArquivo = await premissas();
  const fallback = premissasArquivo.indicadores || {};
  const retrato = await retratoMercado();

  const final = {};
  const meta = {};
  let degradado = false;

  // As chaves saem do retrato; o fallback só entra para as que o retrato
  // conhece mas não conseguiu preencher.
  for (const [chave, campo] of Object.entries(retrato.campos || {})) {
    if (campo.valor !== null && campo.valor !== undefined) {
      final[chave] = campo.valor;
      meta[chave] = {
        rotulo: campo.rotulo || chave,
        unidade: campo.unidade || '',
        origem: campo.fonte || 'fonte oficial',
        heranca: campo.origem !== 'fonte oficial' ? campo.origem : null,
        referencia_fonte: campo.referencia_fonte,
        referencia: campo.referencia,
      };
    } else if (chave in fallback) {
      final[chave] = fallback[chave];
      meta[chave] = {
        rotulo: campo.rotulo || chave,
        unidade: campo.unidade || '',
        origem: 'fallback de premissas.json',
        referencia: premissasArquivo.atualizado_em,
        detalhe: campo.detalhe || 'fonte indisponível',
      };
      degradado = true;
    }
    // Campo sem valor e sem fallback fica de fora: melhor ausente que inventado.
  }

  // Retrato ilegível: sobra o que premissas.json souber, todo marcado.
  if (!Object.keys(final).length) {
    for (const [chave, valor] of Object.entries(fallback)) {
      if (chave.startsWith('_')) continue;
      final[chave] = valor;
      meta[chave] = {
        rotulo: chave,
        unidade: '',
        origem: 'fallback de premissas.json',
        referencia: premissasArquivo.atualizado_em,
        detalhe: retrato._erro || 'retrato de mercado indisponível',
      };
    }
    degradado = true;
  }

  const idade = diasDesde(retrato.coletado_em);
  final._meta = meta;
  final._degradado = degradado;
  final._coletado_em = retrato.coletado_em || null;
  final._idade_dias = idade;
  final._vencido = idade !== null && idade > DIAS_ATE_VENCER;
  return final;
}

/** Manifesto das séries disponíveis: [{id, rotulo, rotulo_curto, ...}]. */
export async function seriesDisponiveis() {
  const retrato = await retratoMercado();
  return retrato.series || [];
}

/** Série histórica mensal real, acumulada a partir do primeiro ponto. */
export async function historico(serieId, meses = 60) {
  const disponiveis = await seriesDisponiveis();
  const entrada = disponiveis.find((s) => s.id === serieId);
  if (!entrada) {
    const ids = disponiveis.map((s) => s.id).join(', ') || 'nenhuma';
    throw new Error(`Série "${serieId}" não está no retrato. Disponíveis: ${ids}`);
  }

  const arquivo = await lerJson(DIR_MERCADO + entrada.arquivo);

  // O mês corrente é publicado parcial pela fonte: incluí-lo desenharia uma
  // queda falsa no último ponto.
  const mesAtual = new Date().toISOString().slice(0, 7);
  const pontos = arquivo.pontos
    .filter((p) => p.data.slice(0, 7) !== mesAtual)
    .slice(-meses);
  if (!pontos.length) throw new Error(`Série "${serieId}" sem meses fechados disponíveis`);

  let fator = 1;
  const acumulado = pontos.map((p) => {
    fator *= 1 + p.valor / 100;
    return {
      data: p.data,
      mensal_pct: p.valor,
      acumulado_pct: Math.round((fator - 1) * 1e6) / 1e4,
      indice_100: Math.round(fator * 1e6) / 1e4,
    };
  });

  return {
    ...entrada,
    coletado_em: arquivo.coletado_em,
    meses_disponiveis: arquivo.pontos.length,
    pontos: acumulado,
    acumulado_total_pct: acumulado[acumulado.length - 1].acumulado_pct,
  };
}

/* ---------------------------------------------------------- fundos (FII/ETF) */
/** Manifesto dos fundos coletados: [{ticker, tipo, rotulo, arquivo, ...}]. */
export async function fundosDisponiveis() {
  const retrato = await retratoMercado();
  return retrato.fundos || [];
}

/** Histórico + resumo de um conjunto de tickers, só o que `app/nucleo/` precisa
    para resolver 'fundo_fii' e 'etf_historico'. Pede só os tickers dos ativos
    selecionados - a página não carrega os 13 arquivos se o usuário marcou 2. */
export async function fundos(tickers) {
  const unicos = [...new Set(tickers)];
  const entradas = await Promise.all(unicos.map(async (ticker) => {
    try {
      return [ticker, await lerJson(DIR_FUNDOS + `${ticker}.json`)];
    } catch (erro) {
      // Sem arquivo coletado ainda: o resolver do núcleo é quem decide o que
      // fazer (hoje, o ativo cai em `erros` da comparação com uma mensagem
      // clara) - não inventamos um valor aqui.
      return [ticker, null];
    }
  }));
  return Object.fromEntries(entradas.filter(([, valor]) => valor !== null));
}

File: backend/dados.py
"""Acesso aos arquivos JSON de catalogo e premissas, e a fusao deles com os
indicadores vivos do Banco Central.

Regra de ouro do projeto: numero de fonte oficial sempre vence o fallback do
arquivo, e a origem de cada numero viaja junto com ele ate a tela.
"""
from __future__ import annotations

import json
from datetime import date
from pathlib import Path

from .fontes import bcb_sgs

RAIZ = Path(__file__).resolve().parents[1]
CAMINHO_ATIVOS = RAIZ / "dados" / "catalogo" / "ativos.json"
CAMINHO_PREMISSAS = RAIZ / "dados" / "premissas" / "premissas.json"


def _ler(caminho: Path) -> dict:
    return json.loads(caminho.read_text(encoding="utf-8"))


def catalogo() -> dict:
    return _ler(CAMINHO_ATIVOS)


def ativos_por_id() -> dict:
    return {a["id"]: a for a in catalogo()["ativos"]}


def premissas() -> dict:
    return _ler(CAMINHO_PREMISSAS)


def indicadores(forcar_atualizacao: bool = False) -> dict:
    """Indicadores prontos para o motor, com a procedencia de cada campo.

    Formato: {'cdi_aa': 13.9, ..., '_origem': {'cdi_aa': {...}}, '_degradado': bool}
    """
    base = premissas()["indicadores"]
    ttl = 0 if forcar_atualizacao else bcb_sgs.TTL_PADRAO
    vivos = bcb_sgs.indicadores(ttl=ttl)

    final = {}
    procedencia = {}
    degradado = False

    mapa = {
        "cdi_aa": "cdi_aa",
        "selic_meta_aa": "selic_meta_aa",
        "ipca_12m": "ipca_12m",
        "poupanca_am": "poupanca_am",
    }
    for chave, campo in mapa.items():
        vivo = vivos["campos"].get(campo, {})
        if vivo.get("valor") is not None:
            final[chave] = vivo["valor"]
            procedencia[chave] = {
                "origem": "Banco Central (SGS)",
                "serie_sgs": vivo.get("serie_sgs"),
                "referencia": vivo.get("referencia"),
            }
        else:
            final[chave] = base[chave]
            procedencia[chave] = {
                "origem": "fallback de premissas.json",
                "referencia": premissas()["atualizado_em"],
                "detalhe": vivo.get("detalhe", "fonte indisponivel"),
            }
            degradado = True

    final["tr_am"] = base.get("tr_am", 0.0)
    final["_origem"] = procedencia
    final["_degradado"] = degradado
    final["_consultado_em"] = vivos["consultado_em"]
    return final


def historico(serie_nome: str, meses: int = 60) -> dict:
    """Serie historica mensal real, para a linha do tempo do simulador."""
    disponiveis = {
        "cdi": ("cdi_mensal", "CDI acumulado no mes", 4391),
        "selic": ("selic_mensal", "Selic acumulada no mes", 4390),
        "ipca": ("ipca_mensal", "IPCA mensal", 433),
        "poupanca": ("poupanca_mensal", "Rendimento da poupanca", 196),
    }
    if serie_nome not in disponiveis:
        raise ValueError(
            f"Serie '{serie_nome}' indisponivel. Opcoes: {', '.join(disponiveis)}"
        )
    chave, rotulo, codigo = disponiveis[serie_nome]
    # Pedimos um mes a mais porque o mes corrente vem parcial e sera descartado.
    pontos = bcb_sgs.serie(chave, ultimos=max(2, min(meses + 1, 360)))

    mes_atual = date.today().strftime("%Y-%m")
    pontos = [p for p in pontos if p["data"][:7] != mes_atual][-meses:]
    if not pontos:
        raise ValueError(f"Serie '{serie_nome}' sem meses fechados disponiveis")

    fator = 1.0
    acumulado = []
    for ponto in pontos:
        fator *= 1 + ponto["valor"] / 100
        acumulado.append({
            "data": ponto["data"],
            "mensal_pct": ponto["valor"],
            "acumulado_pct": round((fator - 1) * 100, 4),
            "indice_100": round(fator * 100, 4),
        })

    return {
        "serie": serie_nome,
        "rotulo": rotulo,
        "fonte": "Banco Central do Brasil - SGS",
        "serie_sgs": codigo,
        "pontos": acumulado,
        "acumulado_total_pct": acumulado[-1]["acumulado_pct"] if acumulado else 0.0,
    }
